src/toil/job.py

```diff
@@ -1722,3 +1721,185 @@ def convertPromises(kwargs):
+            path = job.fileStore.readGlobalFile(ref_id, os.path.join(work_dir, 'ref.fasta')
+            parameters = ['faidx', path]
+            docker_call(job, tool='quay.io/ucgc_cgl/samtools', work_dir=work_dir, parameters=parameters)
+
+    :param toil.Job.job job: The Job instance for the calling function.
+    :param str tool: Name of the Docker image to be used (e.g. quay.io/ucsc_cgl/samtools).
+    :param list[str] parameters: Command line arguments to be passed to the tool.
+    :param str work_dir: Directory to mount into the container via `-v`. Destination convention is /data
+    :param list[str,] docker_parameters: Parameters to pass to Docker. Default parameters are `--rm`,
+            `--log-driver none`, and the mountpoint `-v work_dir:/data` where /data is the destination convention.
+             These defaults are removed if docker_parmaters is passed, so be sure to pass them if they are desired.
+    :param file outfile: Pipe output of Docker call to file handle
+    :param bool check_output: When True, this function returns docker's output.
+    :param int defer: What action should be taken on the container upon job completion?
+           docker_call.FORGO (0) will leave the container untouched.
+           docker_call.STOP (1) will attempt to stop the container with `docker stop` (useful for
+           debugging).
+           docker_call.RM (2) will stop the container and then forcefully remove it from the system
+           using `docker rm -f`.
+           The default value is None and that shadows docker_call.FORGO, unless --rm is being passed in.
+    """
+    docker_call.FORGO = 0
+    docker_call.STOP = 1
+    docker_call.RM = 2
+
+    if parameters is None:
+        parameters = []
+
+    # Setup the outgoing subprocess call for docker
+    base_docker_call = ['docker', 'run']
+    if docker_parameters:
+        base_docker_call += docker_parameters
+    else:
+        base_docker_call += ['--rm', '--log-driver', 'none', '-v',
+                             os.path.abspath(work_dir) + ':/data']
+
+    # Ensure the user has passed a valid value for defer
+    require(defer in (None, docker_call.FORGO, docker_call.STOP, docker_call.RM),
+            'Please provide a valid value for defer.')
+
+    # Get container name which is needed for _docker_kill
+    try:
+        if any('--name=' in x for x in base_docker_call):
+            container_name = [x.split('=')[1] for x in base_docker_call if '--name=' in x][0]
+        else:
+            container_name = base_docker_call[base_docker_call.index('--name') + 1]
+    except ValueError:
+        container_name = _get_container_name(job)
+        base_docker_call.extend(['--name', container_name])
+    except IndexError:
+        raise RuntimeError("Couldn't parse Docker's `--name=` option, check parameters: " + str(base_docker_call))
+
+    # Defer the container on-exit action
+    if '--rm' in base_docker_call and defer is None:
```

Comment:
I see a couple places this would cause a `CalledProcessError` in `_dockerKill`, but whether it matters depends on if deferred functions failing affect the pipeline at all. If they don't, then there's no side-effects. 

Since it's likely unintended behavior on the user's behalf, it probably makes sense to either warn them if STOP / FORGO and `--rm` are both being used, raise an error, or remove `--rm` if either STOP / FORGO are passed.